op_crates/websocket/02_websocketstream.js

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+"use strict";
+
+((window) => {
+  const core = window.Deno.core;
+
+  /** @template T */
+  class Deferred {
+    /** @type {Promise<T>} */
+    #promise;
+    /** @type {(reject?: any) => void} */
+    #reject;
+    /** @type {(value: T | PromiseLike<T>) => void} */
+    #resolve;
+    /** @type {"pending" | "fulfilled"} */
+    #state = "pending";
+
+    constructor() {
+      this.#promise = new Promise((resolve, reject) => {
+        this.#resolve = resolve;
+        this.#reject = reject;
+      });
+    }
+
+    /** @returns {Promise<T>} */
+    get promise() {
+      return this.#promise;
+    }
+
+    /** @returns {"pending" | "fulfilled"} */
+    get state() {
+      return this.#state;
+    }
+
+    /** @param {any=} reason */
+    reject(reason) {
+      // already settled promises are a no-op
+      if (this.#state !== "pending") {
+        return;
+      }
+      this.#state = "fulfilled";
+      this.#reject(reason);
+    }
+
+    /** @param {T | PromiseLike<T>} value */
+    resolve(value) {
+      // already settled promises are a no-op
+      if (this.#state !== "pending") {
+        return;
+      }
+      this.#state = "fulfilled";
+      this.#resolve(value);
+    }
+  }
+
+  /**
+   * Tries to close the resource (and ignores BadResource errors).
+   * @param {number} rid
+   */
+  function tryClose(rid) {
+    try {
+      core.close(rid);
+    } catch (err) {
+      // Ignore error if the socket has already been closed.
+      if (!(err instanceof Deno.errors.BadResource)) throw err;
+    }
+  }
+
+  class WebSocketStream {
+    #rid;
+
+    #url;
+    get url() {
+      return this.#url;
+    }
+
+    constructor(url, options) {
+      requiredArguments("WebSocket", arguments.length, 1);
+
+      const wsURL = new URL(url);
+
+      if (wsURL.protocol !== "ws:" && wsURL.protocol !== "wss:") {
+        throw new DOMException(
+          "Only ws & wss schemes are allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      if (wsURL.hash !== "" || wsURL.href.endsWith("#")) {
+        throw new DOMException(
+          "Fragments are not allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      this.#url = wsURL.href;
+
+      core.opSync("op_ws_check_permission", this.#url);
+
+      if (
+        options?.protocols?.some((x) =>
```

Comment:
Slight optimization:
```javascript
if (options?.protocols?.length > 1 && options.protocols.some((x, i) => options.protocols.indexOf(x, i + 1) !== -1))
```

Reply:
Might be cleaner and possibly faster (though I suspect not) to just construct a `Set()` of the protocols and then compare cardinality/length. I would opt for the cleaner route in absence of benchmarks or being in a known hotpath.

Reply:
This is an identical copy of the code used in the normal WebSocket client. Such optimizations would make more sense in their own PR, or most likely together with when we make WebSocket more spec compliant.

Reply:
Agreed